### PR TITLE
Implement new public API for verifying UCANs

### PR DESCRIPTION
**Pull request by matheus23**
This PR mainly implements two big APIs:
* the `delegationChains` function in `attenuation.ts` that gives you proof chains for all capabilities that are granted by your UCAN.
* the `verify` function that uses the above :point_up_2: to verify that a UCAN is valid & has all expected capabilities. From the discussion https://github.com/ucan-wg/ts-ucan/discussions/75

Also
* I've removed `hasCapability` as that should be covered either be the new `verify` function, or using `capabilityCanBeDelegated` on one of the `DelegationChain`s from using `delegationChains`.

### CODE REVIEW

**Comment by matheus23** [comment]
I've switch the base branch of this from the other PR's branch (#78) to `main`, since I think this should just be merged in one piece.

**Comment by therealjeffg** [comment]
Closes #21

File: src/verify.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ * @param ucan
+ *  a UCAN to verify for invocation in JWT format. (starts with 'eyJ...' and has two '.' in it)
+ *
+ * @param audience
+ *  the DID of the callee of this function. The expected audience of the outermost level of the UCAN.
+ *  NOTE: This DID should not be hardcoded in production calls to this function.
+ *
+ * @param isRevoked
+ *  an async predicate on UCANs to figure out whether they've been revoked or not.
+ *  Usually that means checking whether the hash of the UCAN is in a list of revoked UCANs.
```
Comment (icidasset) [inline src/verify.ts:27]: Should we default this to `() => false`?

**Review by icidasset — COMMENTED** [review]
The new capability checking, aka. delegation semantics is great! So much better than before 🎉 Probably should have some other people with better UCAN knowledge review that part as well, but it looked 👌 to me. Not having the `Chained` type also seems like a good idea 👍 

I did have some feedback and questions on the `verify` function.

File: src/verify.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ * @throws TypeError if the passed arguments don't match what is expected
+ */
+export async function verify(
+  ucan: string,
+  audience: string,
+  isRevoked: (ucan: Ucan) => Promise<boolean>,
+  requiredCapabilities: { capability: Capability; rootIssuer: string }[],
+  semantics: DelegationSemantics = equalCanDelegate,
+  checkFacts: (facts: Fact[]) => boolean = () => true,
+): Promise<Result<Verification[], Error[]>> {
```
[the remark this answers is not in the document]
Comment (matheus23) [reply src/verify.ts:52]: I thought about that while writing the readme. I think you're right. Also you may mix up the first and second parameter. Both are strings, so you have no idea you're mixing stuff up, it goes through type checks and just fails weirdly at runtime.

Will do!

File: src/verify.ts (hunk 1 above)
Comment (matheus23) [reply src/verify.ts:27]: My thinking was "that's too dangerous". But actually defaulting this to not have revocation isn't that dangerous after all. Hm. Maybe we can do that.

File: src/verify.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ *
+ * @throws TypeError if the passed arguments don't match what is expected
+ */
+export async function verify(
+  ucan: string,
+  audience: string,
+  isRevoked: (ucan: Ucan) => Promise<boolean>,
+  requiredCapabilities: { capability: Capability; rootIssuer: string }[],
+  semantics: DelegationSemantics = equalCanDelegate,
+  checkFacts: (facts: Fact[]) => boolean = () => true,
```
[the remark this answers is not in the document]
Comment (icidasset) [reply src/verify.ts]: Oh I see now, I didn't squint hard enough. Didn't see the `=` with all those `=>` 😄  My bad!

**Review by icidasset — APPROVED** [review]

File: src/verify.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (matheus23) [reply src/verify.ts]: Yeah, that really was an operator soup!
It's better after the most recent commits now. The default isn't set on the same line as the type signature anymore :)

**Pull request merged**